test/test_constraints.py

```diff
@@ -69,3 +70,78 @@ def test_unknown_labels_not_in_chunks(num_chunks, chunk_size):
                               random_state=SEED)
 
   assert np.all(chunks[labels < 0] < 0)
+
+
+def test_generate_knntriplets():
+  k = 1
+  X = np.array([[0, 0],
+               [1, 1],
+               [2, 2],
+               [3, 3],
+               [4, 4],
+               [5, 5],
+               [6, 6],
+               [7, 7],
+               [8, 8]])
+  y = np.array([1, 1, 1, 2, 2, 2, 3, 3, 3])
+  T_test = np.array([[0, 1, 3],
+                    [1, 0, 3],
+                    [2, 1, 3],
+                    [3, 4, 2],
+                    [4, 3, 2],
+                    [5, 4, 6],
+                    [6, 7, 5],
+                    [7, 6, 5],
+                    [8, 7, 5]])
+  T = Constraints(y).generate_knntriplets(X, k, k)
+
+  assert np.array_equal(T, T_test)
```

Comment:
this example is not ideal as there are ties (e.g., data point 1 is at equal distance from 0 and from 2), so this equality check assumes a specific tie breaking strategy. It would be better to modify the example so that it does not involve ties.

this check also assumes a particular ordering in the triplet generation. it is probably better to sort the rows so that the check is independent of the ordering

Reply:
Totally agree with you. Fixed these two issues in the las commit!